### PR TITLE
refactor: use pgxpool.Pool instead of database/sql for postgresql

### PR DESCRIPTION
**Pull request by aboudreault**
This PR is a refactor of the postgres persistence layer. Changes:

* Use the `pgxpool.Pool` interface to execute query instead of the built-in `database/sql`
* Add basic pooling options to the postgres config options.
* Make the postgres Pool plugable under the hood.

### CODE REVIEW

File: internal/persistence/postgres/postgres.go
Comment (aboudreault) [inline internal/persistence/postgres/postgres.go:116]: With `database/sql`, a transaction will be automatically roll backed on context cancellation. This is not something we have by default with `pgx`. Although we don't have any test that is testing this scenario, I think it would be a good to do the same thing here, wdyt?

File: internal/persistence/postgres/pool.go
Comment (aboudreault) [inline internal/persistence/postgres/pool.go:48]: Currently, this is a bit falsy. The default are only set in the config yaml stuff...

File: internal/persistence/postgres/pool.go
Comment (aboudreault) [inline internal/persistence/postgres/pool.go:68]: Building a new PoolOpts  using `PoolOpts{}` will give us an invalid default object, due to the zero-value of these types (see PoolOpts.Validate()). This is not ideal. Should I create a proper `newPooOpts()` to avoid this? (not sure yet how this would work since we also use `PostgresOpts{}` directly.

File: internal/persistence/postgres/pool.go
Comment (aboudreault) [inline internal/persistence/postgres/pool.go:153]: use `RegisterPool()` and handle the error with a simple panic.

**Comment by aboudreault** [comment]
@Kong/koko-engineering this is ready for a first review, if anyone has some free cycles

File: internal/config/db.go
Comment (omegabytes) [inline internal/config/db.go:82]: How did we choose these as default values?

Comment (aboudreault) [reply internal/config/db.go:82]: these defaults come from koko if they were existing, otherwise pgx itself.

File: internal/persistence/postgres/pool.go
Comment (aboudreault) [reply internal/persistence/postgres/pool.go:68]: yeah, I wrote this comment prior the addition of the pool validation function: https://github.com/Kong/koko/pull/502/files#diff-d18f5e36820a1a2d869d4c377f9d3bc3ee07bd24391f4c43ec7e7a6b80e1b5f9R71

so with the default yaml settings being applied at config loading.. and also koko calling this Validate function, the normal path is fine.

File: internal/persistence/postgres/options.go
Comment (GGabriele) [inline internal/persistence/postgres/options.go:57]:
```suggestion
	// Pool defines configuration for connections pooling.
```

File: internal/persistence/postgres/pool.go
Comment (GGabriele) [inline internal/persistence/postgres/pool.go:59]:
```suggestion
	// closed by the health check. Default to `persistence.MaxConnIdleTime`.
```

File: internal/persistence/postgres/pool_test.go
Comment (GGabriele) [inline internal/persistence/postgres/pool_test.go:75]: Maybe add a test case for an already registered pool?

Comment (aboudreault) [reply internal/persistence/postgres/pool_test.go:75]: +1, will do

File: internal/persistence/postgres/pool.go
Comment (tjasko) [inline internal/persistence/postgres/pool.go:122]: Let's use a context with a hard-coded timeout of like 15s or so, because we wouldn't want this to hang in the event it can't establish the connection or ping in a decent timeframe.

File: internal/persistence/postgres/pool.go
Comment (tjasko) [inline internal/persistence/postgres/pool.go:137]: Why is this happening? `pgxpool.ConnectConfig()` will already establish the connection, so this ping seems unnecessary, no?

File: internal/persistence/postgres/pool.go
Comment (tjasko) [inline internal/persistence/postgres/pool.go:153]: I don't understand why we're doing this. Can we throw some documentation on why we need to store all the registered pools? I imagine it's for the healthcheck stuff to ensure we can easily look up all pools at once, but perhaps I'm wrong.

File: internal/persistence/postgres/pool.go
Comment (tjasko) [inline internal/persistence/postgres/pool.go:67]: How will this be used, as it's currently unused?

File: internal/config/db.go
Comment (tjasko) [inline internal/config/db.go:81]: This should probably be disabled by default, right?

File: internal/persistence/postgres/tx.go
Comment (tjasko) [inline internal/persistence/postgres/tx.go:10]: Thanks for pulling that out like I did with MySQL, def helps keep the code more readable.

**Comment by tjasko** [comment]
Great work here @aboudreault 😄

File: internal/persistence/postgres/postgres.go
Comment (aboudreault) [reply internal/persistence/postgres/postgres.go:116]: ACK: thanks!

File: internal/persistence/postgres/pool.go
Comment (aboudreault) [reply internal/persistence/postgres/pool.go:19]: since `*pgxpool.Pool` is a type, I cannot use it as the abstraction type/interface for my different pool implementations. e.g. My custom aurora pool implementation is returning a `*AuroraPool`. I need a way that our postgres persistence layer can rely on and just call the query execution functions. does that make sense or I am missing something?

File: internal/persistence/postgres/pool.go
Comment (aboudreault) [reply internal/persistence/postgres/pool.go:67]: currently, this is used in the aurora impl to determine what kind of healthcheck validator is required (read vs write). However, we could remove that if we had a generic `PoolOpts.params` that we can pass thru.

File: internal/persistence/postgres/pool.go
Comment (aboudreault) [reply internal/persistence/postgres/pool.go:68]: will fix!

File: internal/persistence/postgres/pool.go
Comment (aboudreault) [reply internal/persistence/postgres/pool.go:93]: indeed, +1!!

File: internal/persistence/postgres/pool.go
Comment (aboudreault) [reply internal/persistence/postgres/pool.go:121]: will fix!

File: internal/persistence/postgres/pool.go
Comment (aboudreault) [reply internal/persistence/postgres/pool.go:122]: good point, will fix!

File: internal/persistence/postgres/pool.go
Comment (aboudreault) [reply internal/persistence/postgres/pool.go:137]: umm, yeah let's remove this.

File: internal/persistence/postgres/pool.go
Comment (aboudreault) [reply internal/persistence/postgres/pool.go:143]: +1, will fix

File: internal/persistence/postgres/pool.go
Comment (aboudreault) [reply internal/persistence/postgres/pool.go:153]: not what what you mean, it's only to be able to select the proper pool implementation by name from the configuration. e.g:
`  postgres:
    hostname: localhost
    port: 5432
    db_name: koko
    user: koko
    password: koko
    pool:
      name: aurora
      health_check_period: 10s
`

so a `pool_name:pool_init_func` registration.

File: internal/persistence/postgres/postgres.go
Comment (aboudreault) [reply internal/persistence/postgres/postgres.go:86]: totally!

**Review by aboudreault — COMMENTED** [review]
Thanks Taylor for all your comments! I responded and will fix a bunch of them tomorrow morning.

File: internal/persistence/postgres/pool_test.go
Comment (aboudreault) [reply internal/persistence/postgres/pool_test.go:67]: done

File: internal/persistence/postgres/postgres.go
Comment (aboudreault) [reply internal/persistence/postgres/postgres.go]: done

File: internal/config/config.go
Comment (rajkong) [inline internal/config/config.go:33]: Do we need this?

Comment (aboudreault) [reply internal/config/config.go:33]: not at all!! will remove

File: internal/config/config_test.go
Comment (aboudreault) [reply internal/config/config_test.go]: done

File: internal/config/config.go
Comment (aboudreault) [reply internal/config/config.go]: done

File: internal/persistence/driver.go
Comment (rajkong) [inline internal/persistence/driver.go:10]: Will driver.go change when users want to add implementations?

File: internal/persistence/postgres/pool.go
Comment (rajkong) [reply internal/persistence/postgres/pool.go:68]: I am not sure about the value of the ReadOnly parameter as well. Perhaps this is something to talk about internally?

Comment (aboudreault) [reply internal/persistence/postgres/pool.go:68]: sure Rajan. Here is how the ro pool is currently created: https://github.com/Kong/koko/pull/502/files#diff-f39571bb34baebee68e4a03486bb984084067ec6ac049be19f648184888d2494R63-R71

feel free to ping me on on Slack and/or suggest an alternative way to do this. we would also add a readOnly argument to ` newPostgresPool` but not sure this is cleaner anyway.

File: internal/persistence/driver.go
Comment (aboudreault) [reply internal/persistence/driver.go:10]: that `SQLOpenFunc` is not used for pool initialization.. I should add something in its docs. maybe you can elaborate on your thoughts about it offline.

File: internal/persistence/postgres/pool.go
Comment (aboudreault) [reply internal/persistence/postgres/pool.go:68]: this was discussed offline with Rajan and we agreed it was indeed the simpler way to accomplish what we need in the custom pool.

Comment (rajkong) [reply internal/persistence/postgres/pool.go:68]: Discussed and the approach is ok

**Review by rajkong — COMMENTED** [review]
Looks good to me. Waiting for other comments to be resolved before approval

File: internal/config/db.go
Comment (tjasko) [reply internal/config/db.go:76]: Ah, got it, wasn't thinking there could be multiple pool implementations (thought that wouldn't been done differently).

File: internal/config/db.go
Comment (tjasko) [reply internal/config/db.go]: Err, I linked to the wrong lib lol. Sorry, the library we're using for config unmarshalling also supports this:
https://github.com/ilyakaznacheev/cleanenv

We just need to change the type to `time.Duration`, and it'll just work as expected.

File: internal/config/db.go
Comment (tjasko) [reply internal/config/db.go]: Yeah that's perfectly fine to cast when appropriate (as after all, pgx is a third-party lib). Thanks!

File: internal/persistence/postgres/pool.go
Comment (tjasko) [reply internal/persistence/postgres/pool.go:133]: Cool, thanks!

File: internal/persistence/postgres/pool.go
Comment (tjasko) [reply internal/persistence/postgres/pool.go]: Err, yeah you can forget this, I see what it's doing now. I should have deleted this comment.

File: internal/config/db.go
Comment (tjasko) [reply internal/config/db.go]: Ohh, completely sidestepped on the fact that the pgx pool implementation implements its own health check too... derp, mind me.

**Comment by tjasko** [comment]
I dropped the ball on this @aboudreault, I apologize for the delay getting back to you. Other than the below comment, all looks good!
https://github.com/Kong/koko/pull/502#discussion_r1037610001

File: internal/config/db.go
Comment (aboudreault) [reply internal/config/db.go]: ah! no idea why I thought Duration was supported in cleanenv without even looking :facepalm: ... maybe because of how `queryTimeout` was handled. fixed!

**Review by tjasko — APPROVED** [review]

**Pull request merged**
